# Task
Optimize team selection (Team A vs. Team B) for a 3805-mile trip by defining driver data (rates, speeds, daily hour limits), solving a multi-day optimization problem for each team using Gurobi to minimize cost while adhering to individual driver and 18-hour team daily limits, and then selecting the team with the lowest cost, or fastest time if costs are equal, and provide a clear explanation.

## Define Driver Data and Constraints

### Subtask:
Store the detailed driver information (rate, speed, and specific daily hour limits) for each driver type (Pro, Amateur) within Team A and Team B. Also, define the overall team daily driving constraint of 18 hours and the target distance of 3805 miles.


**Reasoning**:
To review the current values of the variables as requested by the instructions, I will print `team_A_data`, `team_B_data`, `team_driving_constraint`, and `distance_to_cover`.



In [2]:
print(f"Team A Data: {team_A_data}")
print(f"Team B Data: {team_B_data}")
print(f"Team Driving Constraint: {team_driving_constraint} hours/day")
print(f"Distance to Cover: {distance_to_cover} miles")

Team A Data: {'Pro': {'rate': 30, 'speed': 60, 'hours_per_day': 8}, 'Amateur': {'rate': 25, 'speed': 45, 'hours_per_day': 14}}
Team B Data: {'Pro': {'rate': 32, 'speed': 65, 'hours_per_day': 9}, 'Amateur': {'rate': 23, 'speed': 40, 'hours_per_day': 13}}
Team Driving Constraint: 18 hours/day
Distance to Cover: 3805 miles


## Solve Optimization for Team A

### Subtask:
Using Gurobi, formulate and solve an optimization problem for Team A. This will involve defining decision variables for the total hours each Team A driver works over the entire trip and the number of days the trip takes. Constraints will include individual driver daily hour limits, the total team's 18-hour daily limit, and the total distance (3805 miles) to be covered. The objective will be to minimize Team A's total cost. After solving, record Team A's minimum cost, total driving time, and number of days.


**Reasoning**:
I will import the Gurobi library and define the optimization model for Team A, including decision variables, constraints, and the objective function as outlined in the instructions. After optimizing, I will extract and store the relevant results.



In [5]:
# Install Gurobi
!pip install gurobipy

# Import the Gurobi Package
import gurobipy as gp
from gurobipy import GRB

# Create a new model
m = gp.Model("TeamA_Optimization")

# Define decision variables
x_pro_A = m.addVar(lb=0, vtype=GRB.CONTINUOUS, name="x_pro_A") # Total hours for Pro driver A
x_amateur_A = m.addVar(lb=0, vtype=GRB.CONTINUOUS, name="x_amateur_A") # Total hours for Amateur driver A
num_days_A = m.addVar(lb=0, vtype=GRB.CONTINUOUS, name="num_days_A") # Total number of days for Team A

# Constraints
# Individual driver daily hour limits
m.addConstr(x_pro_A <= team_A_data['Pro']['hours_per_day'] * num_days_A, "Pro_A_Daily_Limit")
m.addConstr(x_amateur_A <= team_A_data['Amateur']['hours_per_day'] * num_days_A, "Amateur_A_Daily_Limit")

# Team daily driving constraint
m.addConstr(x_pro_A + x_amateur_A <= team_driving_constraint * num_days_A, "Team_A_Daily_Limit")

# Total distance covered
m.addConstr(x_pro_A * team_A_data['Pro']['speed'] + \
            x_amateur_A * team_A_data['Amateur']['speed'] >= distance_to_cover, "Total_Distance_A")

# Set objective function: Minimize total cost
m.setObjective(x_pro_A * team_A_data['Pro']['rate'] + \
               x_amateur_A * team_A_data['Amateur']['rate'], GRB.MINIMIZE)

# Optimize the model
m.optimize()

# Retrieve and store results
if m.status == GRB.OPTIMAL:
    team_a_cost = m.objVal
    team_a_time = x_pro_A.x + x_amateur_A.x
    team_a_num_days = num_days_A.x
    print(f"Team A Optimal Cost: {team_a_cost:.2f}")
    print(f"Team A Total Driving Time: {team_a_time:.2f} hours")
    print(f"Team A Number of Days: {team_a_num_days:.2f}")
else:
    print("No optimal solution found for Team A")
    team_a_cost = float('inf')
    team_a_time = float('inf')
    team_a_num_days = float('inf')


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.8/14.8 MB 73.5 MB/s eta 0:00:00
Restricted license - for non-production use only - expires 2027-11-29
Gurobi Optimizer version 13.0.0 build v13.0.0rc1 (linux64 - "Ubuntu 22.04.4 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 4 rows, 3 columns and 9 nonzeros (Min)
Model fingerprint: 0x6ca8d203
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 6e+01]
  Objective range  [2e+01, 3e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [4e+03, 4e+03]
Presolve removed 4 rows and 3 columns
Presolve time: 0.01s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.9025000e+03   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.902500000e+03
Tea

# Task
## Solve Optimization for Team B

### Subtask:
Using Gurobi, formulate and solve an optimization problem for Team B. This will involve defining decision variables for the total hours each Team B driver works over the entire trip and the number of days the trip takes. Constraints will include individual driver daily hour limits, the total team's 18-hour daily limit, and the total distance (3805 miles) to be covered. The objective will be to minimize Team B's total cost. After solving, record Team B's minimum cost, total driving time, and number of days.

## Solve Optimization for Team B

### Subtask:
Using Gurobi, formulate and solve an optimization problem for Team B. This will involve defining decision variables for the total hours each Team B driver works over the entire trip and the number of days the trip takes. Constraints will include individual driver daily hour limits, the total team's 18-hour daily limit, and the total distance (3805 miles) to be covered. The objective will be to minimize Team B's total cost. After solving, record Team B's minimum cost, total driving time, and number of days.


## Solve Optimization for Team B

### Reasoning:
I will now define and solve the optimization problem for Team B using Gurobi, following the same methodology as for Team A. This involves creating decision variables for the total hours worked by Team B's Pro and Amateur drivers, and for the total number of days for the trip. Constraints will be established for individual driver daily hour limits, the overall team daily driving limit (18 hours), and ensuring the target distance of 3805 miles is covered. The objective will be to minimize Team B's total cost. After solving, I will extract and store the optimal cost, total driving time, and number of days for Team B.

**Reasoning**:
I will implement the Gurobi optimization model for Team B, defining decision variables, constraints, and the objective function as per the subtask instructions. After optimization, I will extract and store the results for Team B's cost, driving time, and number of days.



In [6]:
import gurobipy as gp
from gurobipy import GRB

# Create a new model for Team B
m_B = gp.Model("TeamB_Optimization")

# Define decision variables for Team B
x_pro_B = m_B.addVar(lb=0, vtype=GRB.CONTINUOUS, name="x_pro_B") # Total hours for Pro driver B
x_amateur_B = m_B.addVar(lb=0, vtype=GRB.CONTINUOUS, name="x_amateur_B") # Total hours for Amateur driver B
num_days_B = m_B.addVar(lb=0, vtype=GRB.CONTINUOUS, name="num_days_B") # Total number of days for Team B

# Constraints for Team B
# Individual driver daily hour limits
m_B.addConstr(x_pro_B <= team_B_data['Pro']['hours_per_day'] * num_days_B, "Pro_B_Daily_Limit")
m_B.addConstr(x_amateur_B <= team_B_data['Amateur']['hours_per_day'] * num_days_B, "Amateur_B_Daily_Limit")

# Team daily driving constraint
m_B.addConstr(x_pro_B + x_amateur_B <= team_driving_constraint * num_days_B, "Team_B_Daily_Limit")

# Total distance covered
m_B.addConstr(x_pro_B * team_B_data['Pro']['speed'] + \
            x_amateur_B * team_B_data['Amateur']['speed'] >= distance_to_cover, "Total_Distance_B")

# Set objective function: Minimize total cost for Team B
m_B.setObjective(x_pro_B * team_B_data['Pro']['rate'] + \
               x_amateur_B * team_B_data['Amateur']['rate'], GRB.MINIMIZE)

# Optimize the model
m_B.optimize()

# Retrieve and store results for Team B
if m_B.status == GRB.OPTIMAL:
    team_b_cost = m_B.objVal
    team_b_time = x_pro_B.x + x_amateur_B.x
    team_b_num_days = num_days_B.x
    print(f"Team B Optimal Cost: {team_b_cost:.2f}")
    print(f"Team B Total Driving Time: {team_b_time:.2f} hours")
    print(f"Team B Number of Days: {team_b_num_days:.2f}")
else:
    print("No optimal solution found for Team B")
    team_b_cost = float('inf')
    team_b_time = float('inf')
    team_b_num_days = float('inf')

Gurobi Optimizer version 13.0.0 build v13.0.0rc1 (linux64 - "Ubuntu 22.04.4 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 4 rows, 3 columns and 9 nonzeros (Min)
Model fingerprint: 0xfc794820
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 6e+01]
  Objective range  [2e+01, 3e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [4e+03, 4e+03]
Presolve removed 4 rows and 3 columns
Presolve time: 0.01s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.8732308e+03   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.873230769e+03
Team B Optimal Cost: 1873.23
Team B Total Driving Time: 58.54 hours
Team B Number of Days: 6.50


## Compare Teams and Select Optimal

### Subtask:
Compare the optimized results from Team A and Team B. The primary criterion will be the lowest total cost to cover 3805 miles. If there's a tie in cost, the team that achieves the distance in the fastest time (lowest total hours) will be selected as the 'best'.


**Reasoning**:
I will compare the optimal costs and times for Team A and Team B, then store the details of the best performing team as specified in the instructions.



In [7]:
if team_a_cost < team_b_cost:
    best_team_name = 'Team A'
    best_team_cost = team_a_cost
    best_team_time = team_a_time
elif team_b_cost < team_a_cost:
    best_team_name = 'Team B'
    best_team_cost = team_b_cost
    best_team_time = team_b_time
else: # Costs are equal, compare times
    if team_a_time < team_b_time:
        best_team_name = 'Team A'
        best_team_cost = team_a_cost
        best_team_time = team_a_time
    else:
        best_team_name = 'Team B'
        best_team_cost = team_b_cost
        best_team_time = team_b_time

print(f"The optimal team is: {best_team_name}")
print(f"Optimal Cost: {best_team_cost:.2f}")
print(f"Total Driving Time: {best_team_time:.2f} hours")

The optimal team is: Team B
Optimal Cost: 1873.23
Total Driving Time: 58.54 hours


## Output Best Team and Reasoning

### Subtask:
Clearly state which team (Team A or Team B) is identified as the 'best' based on the optimization and comparison, providing the specific optimal cost and total time for that team, and explaining the reasoning for its selection based on the given constraints and objectives.


### Best Team Selection

Based on the optimization results, **Team B** has been identified as the optimal choice.

**Reasoning for Selection**:

The primary objective of this optimization problem was to minimize the total cost of the trip. Team B achieved a lower total cost compared to Team A. Specifically:

*   **Team A Optimal Cost**: $1902.50
*   **Team B Optimal Cost**: $1873.23

Since Team B's total cost ($1873.23) is less than Team A's total cost ($1902.50), Team B is selected as the best option. The tie-breaker criterion (shortest total driving time) was not needed in this case, as Team B had a strictly lower cost.

**Selected Team Summary**:

*   **Optimal Team**: Team B
*   **Optimal Cost**: $1873.23
*   **Total Driving Time**: 58.54 hours

## Final Task

### Subtask:
Summarize the entire process, including the definition of detailed driver data and constraints, the setup and solution of multi-day optimization problems for each team using Gurobi, and the final decision logic to select the overall best team based on minimal cost and faster time.


## Summary:

### Data Analysis Key Findings

*   **Team B Optimization**: An optimization model was successfully formulated and solved for Team B using Gurobi. The model minimized total cost while adhering to constraints on individual driver daily hour limits, a team daily driving limit of 18 hours, and covering a total distance of 3805 miles.
    *   **Optimal Cost for Team B**: \$1873.23
    *   **Total Driving Time for Team B**: 58.54 hours
    *   **Number of Days for Team B**: 6.50 days
*   **Team Comparison**: When comparing Team A and Team B, the primary criterion was the lowest total cost.
    *   Team A's optimal cost was \$1902.50.
    *   Team B's optimal cost was \$1873.23.
*   **Optimal Team Selection**: **Team B** was identified as the optimal choice because its total cost of \$1873.23 was lower than Team A's total cost of \$1902.50. The tie-breaker criterion (shortest total driving time) was not needed as Team B had a strictly lower cost.

### Insights or Next Steps

*   The optimization approach effectively identified the most cost-efficient team for the given long-haul trip, demonstrating the value of mathematical programming in logistical planning.
*   Future analysis could explore scenarios with variable speed limits, rest stop requirements, or different driver availability to further refine optimal routing and team selection.


# Task
Optimize team selection between Team A and Team B for a 3805-mile trip. Define driver data, including rates, speeds, and daily hour limits for each driver type (Pro, Amateur) within both teams, ensuring Team A's Pro driver has an 8-hour daily limit. The overall team daily driving constraint is 18 hours. Formulate and solve a multi-day optimization problem for each team using Gurobi to minimize total cost. Finally, select the team with the lowest total cost, or the fastest time if costs are equal, and provide a clear explanation of the selection.

## Define Driver Data and Constraints

### Subtask:
Store the detailed driver information (rate, speed, and specific daily hour limits) for each driver type (Pro, Amateur) within Team A and Team B. Also, define the overall team daily driving constraint of 18 hours and the target distance of 3805 miles.


**Reasoning**:
To fulfill the subtask, I will create dictionaries for Team A and Team B driver data, and define the team driving constraint and distance to cover as specified in the instructions.



In [1]:
team_A_data = {
    'Pro': {'rate': 30, 'speed': 60, 'hours_per_day': 8},
    'Amateur': {'rate': 25, 'speed': 45, 'hours_per_day': 14}
}

team_B_data = {
    'Pro': {'rate': 32, 'speed': 65, 'hours_per_day': 9},
    'Amateur': {'rate': 23, 'speed': 40, 'hours_per_day': 13}
}

team_driving_constraint = 18
distance_to_cover = 3805

print("Driver data and constraints have been defined.")

Driver data and constraints have been defined.
